File: data-structures/list/problems/go/leetcode-238.go
package problems

func productExceptSelf(nums []int) []int {
	n := len(nums)
	ans := make([]int, n)
	ans[0] = 1

	prefix := 1
	for i := 1; i < n; i++ {
		t := nums[i-1] * prefix
		ans[i] = t
		prefix = t
	}

	suffix := 1
	for i := n - 2; i >= 0; i-- {
		t := nums[i+1] * suffix
		ans[i] = t * ans[i]
		suffix = t
	}
	return ans
}

File: data-structures/list/problems/java/leetcode-238.java
class Solution {
    public int[] productExceptSelf(int[] nums) { // [1,2,3,4]
        int n = nums.length;
        int[] ans = new int[n];
        ans[0] = 1;

        int prefix = 1;
        for (int i = 1; i < n; i++) { // [1,1,2,6]
            int temp = prefix * nums[i - 1];
            ans[i] = temp;
            prefix = temp;
        }

        int suffix = 1;
        for (int j = n - 2; j >= 0; j--) {
            int temp = suffix * nums[j + 1];
            ans[j] = temp * ans[j];
            suffix = temp;
        }

        return ans;

    }
}